File: problems/07_trees/0226_invert_binary_tree/solution.dart
// Problem: 0226. Invert Binary Tree
// Difficulty: Easy

class TreeNode {
  int val;
  TreeNode? left, right;
  TreeNode([this.val = 0, this.left, this.right]);
}

TreeNode? fromLevel(List<int?> a) {
  if (a.isEmpty) return null;
  final root = TreeNode(a[0]!);
  final q = <TreeNode>[root];
  var qi = 0, i = 1;
  while (qi < q.length && i < a.length) {
    final n = q[qi++];
    if (i < a.length && a[i] != null) { n.left = TreeNode(a[i]!); q.add(n.left!); }
    i++;
    if (i < a.length && a[i] != null) { n.right = TreeNode(a[i]!); q.add(n.right!); }
    i++;
  }
  return root;
}

List<int?> toLevel(TreeNode? root) {
  if (root == null) return [];
  final out = <int?>[];
  final q = <TreeNode?>[root];
  var i = 0;
  while (i < q.length) {
    final n = q[i++];
    if (n == null) { out.add(null); continue; }
    out.add(n.val);
    q.add(n.left); q.add(n.right);
  }
  while (out.isNotEmpty && out.last == null) out.removeLast();
  return out;
}

TreeNode? invertTree(TreeNode? root) {
  if (root == null) return null;
  final l = invertTree(root.right);
  final r = invertTree(root.left);
  root.left = l; root.right = r;
  return root;
}

void _expect(List<int?> a, List<int?> e, String msg) {
  if (a.length != e.length || !List.generate(a.length, (i) => a[i] == e[i]).every((x) => x)) {
    throw StateError('FAIL $msg: $a vs $e');
  }
}

void main() {
  _expect(toLevel(invertTree(fromLevel([4, 2, 7, 1, 3, 6, 9]))), [4, 7, 2, 9, 6, 3, 1], 'ex1');
  _expect(toLevel(invertTree(fromLevel([2, 1, 3]))), [2, 3, 1], 'ex2');
  _expect(toLevel(invertTree(fromLevel([]))), [], 'empty');
  _expect(toLevel(invertTree(fromLevel([1]))), [1], 'one');
  print('All tests passed!');
}

File: problems/07_trees/0226_invert_binary_tree/solution.swift
// Problem: 0226. Invert Binary Tree
// Difficulty: Easy

import Foundation

class TreeNode {
    var val: Int; var left: TreeNode?; var right: TreeNode?
    init(_ v: Int = 0, _ l: TreeNode? = nil, _ r: TreeNode? = nil) { val = v; left = l; right = r }
}

func fromLevel(_ a: [Int?]) -> TreeNode? {
    if a.isEmpty { return nil }
    let root = TreeNode(a[0]!)
    var q: [TreeNode] = [root]
    var qi = 0, i = 1
    while qi < q.count && i < a.count {
        let n = q[qi]; qi += 1
        if i < a.count, let v = a[i] { n.left = TreeNode(v); q.append(n.left!) }
        i += 1
        if i < a.count, let v = a[i] { n.right = TreeNode(v); q.append(n.right!) }
        i += 1
    }
    return root
}

func toLevel(_ root: TreeNode?) -> [Int?] {
    guard let root = root else { return [] }
    var out: [Int?] = []
    var q: [TreeNode?] = [root]
    var i = 0
    while i < q.count {
        let n = q[i]; i += 1
        if n == nil { out.append(nil); continue }
        out.append(n!.val)
        q.append(n!.left); q.append(n!.right)
    }
    while !out.isEmpty && out.last! == nil { out.removeLast() }
    return out
}

class Solution {
    func invertTree(_ root: TreeNode?) -> TreeNode? {
        // TODO: implement
        return nil
    }
}

func checkLevel(_ a: [Int?], _ e: [Int?], _ msg: String) {
    if a.count != e.count { fatalError("FAIL \(msg): \(a) vs \(e)") }
    for i in 0..<a.count { if a[i] != e[i] { fatalError("FAIL \(msg): \(a) vs \(e)") } }
}

// MARK: - Tests
// let s = Solution()

// Test 1
// let result = s.solve(...)

print("All tests passed!")
